File: code_challenge_solution/java/32 Connect Four.java
```java
import java.util.Scanner;

public class ConnectFour {
    private static final int ROWS = 6;
    private static final int COLUMNS = 7;
    private static final char EMPTY = '-';
    private static final char PLAYER1 = 'X';
    private static final char PLAYER2 = 'O';

    private char[][] board;
    private char currentPlayer;

    public ConnectFour() {
        board = new char[ROWS][COLUMNS];
        currentPlayer = PLAYER1;
        initializeBoard();
    }

    private void initializeBoard() {
        for (int row = 0; row < ROWS; row++) {
            for (int col = 0; col < COLUMNS; col++) {
                board[row][col] = EMPTY;
            }
        }
    }

    public void playGame() {
        Scanner scanner = new Scanner(System.in);
        boolean gameOver = false;
        
        while (!gameOver) {
            System.out.println("Current board:");
            printBoard();
            
            System.out.print("Player " + currentPlayer + ", choose a column (1-" + COLUMNS + "): ");
            int col = scanner.nextInt() - 1;
            
            if (col < 0 || col >= COLUMNS) {
                System.out.println("Invalid column! Please choose a column between 1 and " + COLUMNS + ".");
                continue;
            }
            
            if (!isColumnFull(col)) {
                int row = dropPiece(col);
                if (checkWin(row, col)) {
                    System.out.println("Player " + currentPlayer + " wins!");
                    gameOver = true;
                } else if (isBoardFull()) {
                    System.out.println("It's a draw!");
                    gameOver = true;
                } else {
                    currentPlayer = (currentPlayer == PLAYER1) ? PLAYER2 : PLAYER1;
                }
            } else {
                System.out.println("Column is full! Please choose another column.");
            }
        }
        
        scanner.close();
    }

    private void printBoard() {
        for (int row = 0; row < ROWS; row++) {
            for (int col = 0; col < COLUMNS; col++) {
                System.out.print(board[row][col] + " ");
            }
            System.out.println();
        }
    }

    private boolean isColumnFull(int col) {
        return board[0][col] != EMPTY;
    }

    private int dropPiece(int col) {
        for (int row = ROWS - 1; row >= 0; row--) {
            if (board[row][col] == EMPTY) {
                board[row][col] = currentPlayer;
                return row;
            }
        }
        return -1; // Should never happen if isColumnFull() is called first
    }

    private boolean checkWin(int row, int col) {
        return checkVerticalWin(row, col) || checkHorizontalWin(row, col) || checkDiagonalWin(row, col);
    }

    private boolean checkVerticalWin(int row, int col) {
        int count = 0;
        for (int i = row; i < ROWS && board[i][col] == currentPlayer; i++) {
            count++;
        }
        return count >= 4;
    }

    private boolean checkHorizontalWin(int row, int col) {
        int count = 0;
        for (int i = col - 3; i <= col + 3; i++) {
            if (i >= 0 && i + 3 < COLUMNS &&
                    board[row][i] == currentPlayer &&
                    board[row][i + 1] == currentPlayer &&
                    board[row][i + 2] == currentPlayer &&
                    board[row][i + 3] == currentPlayer) {
                return true;
            }
        }
        return false;
    }

    private boolean checkDiagonalWin(int row, int col) {
        int count = 0;
        for (int i = -3; i <= 3; i++) {
            if (row + i >= 0 && row + i + 3 < ROWS && col + i >= 0 && col + i + 3 < COLUMNS &&
                    board[row + i][col + i] == currentPlayer &&
                    board[row + i + 1][col + i + 1] == currentPlayer &&
                    board[row + i + 2][col + i + 2] == currentPlayer &&
                    board[row + i + 3][col + i + 3] == currentPlayer) {
                return true;
            }
            if (row - i >= 0 && row - i - 3 < ROWS && col + i >= 0 && col + i + 3 < COLUMNS &&
                    board[row - i][col + i] == currentPlayer &&
                    board[row - i - 1][col + i + 1] == currentPlayer &&
                    board[row - i - 2][col + i + 2] == currentPlayer &&
                    board[row - i - 3][col + i + 3] == currentPlayer) {
                return true;
            }
        }
        return false;
    }

    private boolean isBoardFull() {
        for (int col = 0; col < COLUMNS; col++) {
            if (!isColumnFull(col)) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        ConnectFour game = new ConnectFour();
        game.playGame();
    }
}
```

File: code_challenge_solution/c++/32 Connect Four.cpp
#include <iostream>
#include <vector>

// Function to print the game board
void printBoard(const std::vector<std::vector<char>>& board) {
    for (int row = 0; row < 6; row++) {
        for (int col = 0; col < 7; col++) {
            std::cout << board[row][col] << " ";
        }
        std::cout << std::endl;
    }
}

// Function to check if there is a winning move
bool checkWin(const std::vector<std::vector<char>>& board, char player) {
    // Check rows
    for (int row = 0; row < 6; row++) {
        for (int col = 0; col < 4; col++) {
            if (board[row][col] == player && board[row][col+1] == player &&
                board[row][col+2] == player && board[row][col+3] == player) {
                return true;
            }
        }
    }
    
    // Check columns
    for (int row = 0; row < 3; row++) {
        for (int col = 0; col < 7; col++) {
            if (board[row][col] == player && board[row+1][col] == player &&
                board[row+2][col] == player && board[row+3][col] == player) {
                return true;
            }
        }
    }
    
    // Check diagonals (top-left to bottom-right)
    for (int row = 0; row < 3; row++) {
        for (int col = 0; col < 4; col++) {
            if (board[row][col] == player && board[row+1][col+1] == player &&
                board[row+2][col+2] == player && board[row+3][col+3] == player) {
                return true;
            }
        }
    }
    
    // Check diagonals (top-right to bottom-left)
    for (int row = 0; row < 3; row++) {
        for (int col = 3; col < 7; col++) {
            if (board[row][col] == player && board[row+1][col-1] == player &&
                board[row+2][col-2] == player && board[row+3][col-3] == player) {
                return true;
            }
        }
    }
    
    return false;
}

int main() {
    std::vector<std::vector<char>> board(6, std::vector<char>(7, ' '));
    int currentPlayer = 1;
    int moves = 0;
    
    while (true) {
        // Print the board
        printBoard(board);
        
        if (currentPlayer == 1) {
            std::cout << "Player 1's turn (X): ";
        } else {
            std::cout << "Player 2's turn (O): ";
        }
        
        int col;
        std::cin >> col;
        
        // Check if the column is valid and make the move
        if (col >= 0 && col < 7 && board[0][col] == ' ') {
            for (int row = 5; row >= 0; row--) {
                if (board[row][col] == ' ') {
                    board[row][col] = (currentPlayer == 1) ? 'X' : 'O';
                    moves++;
                    break;
                }
            }
            
            // Check if there is a winning move
            if (checkWin(board, (currentPlayer == 1) ? 'X' : 'O')) {
                printBoard(board);
                if (currentPlayer == 1) {
                    std::cout << "Player 1 (X) wins!" << std::endl;
                } else {
                    std::cout << "Player 2 (O) wins!" << std::endl;
                }
                break;
            }
            
            // Check if the game is a draw
            if (moves == 42) {
                printBoard(board);
                std::cout << "It's a draw!" << std::endl;
                break;
            }
            
            // Switch players
            currentPlayer = (currentPlayer == 1) ? 2 : 1;
        } else {
            std::cout << "Invalid move. Please try again." << std::endl;
        }
    }
    
    return 0;
}